### PR TITLE
Handle redirects with new FollowRedirects state

### PR DESCRIPTION
**Pull request by cwaldren-ld**
This PR finishes off the last of the skipped contract tests, which was the 301 & 307 redirect handling.

It adds a new `FollowRedirects` state, which allows the client to follow the `Location` header sent by the server up to a maximum limit. 

Users can configure the limit using `ClientBuilder::redirect_limit`, but they cannot remove it entirely.

I've also added some new SSE tests to verify a couple edge cases (like missing Location header), which I'll put up for review on that repo.

https://github.com/launchdarkly/sse-contract-tests/pull/17

### CODE REVIEW

File: eventsource-client/src/client.rs [hunk 1]
```diff
@@ -426,6 +490,35 @@ where
                             .set(State::WaitingToReconnect(delay(duration, "retrying")))
                     }
                 },
```
Comment (cwaldren-ld) [inline eventsource-client/src/client.rs:492]: I wasn't sure if usage of `std::io::Error` here constitutes an abuse of that type.

File: eventsource-client/src/client.rs [hunk 2]
```diff
@@ -473,6 +547,23 @@ where
     }
 }
 
+fn uri_from_header(maybe_header: &Option<HeaderValue>) -> Result<Uri> {
+    let header = maybe_header.as_ref().ok_or_else(|| {
+        Error::MalformedLocationHeader(Box::new(std::io::Error::new(
+            ErrorKind::NotFound,
+            "missing Location header",
+        )))
```
Comment (cwaldren-ld) [inline eventsource-client/src/client.rs:555]: Not sure if this is considered an abuse of `std::io::Error`

File: eventsource-client/src/client.rs [hunk 3]
```diff
@@ -137,6 +144,15 @@ impl ClientBuilder {
...
     }
 
+    /// Customize the client's following behavior when served a redirect.
+    /// It is not possible to specify unlimited redirects: `None` informs the client that
+    /// [`DEFAULT_REDIRECT_LIMIT`] should be used.
+    /// To disable redirect following, pass `Some(0)`.
+    pub fn redirect_limit(mut self, limit: Option<u32>) -> ClientBuilder {
+        self.max_redirects = limit.unwrap_or(DEFAULT_REDIRECT_LIMIT);
+        self
+    }
```
[the remark this answers is not in the document]
Comment (cwaldren-ld) [reply eventsource-client/src/client.rs:154]: I like that, although I was trying to support the use case of removing the custom limit (via None). Perhaps that is not a realistic use-case though.
```rust
// Set a general limit of 10 redirects
builder.redirect_limit(Some(10))

if config.using_some_feature() {
  // Nevermind, use the default
  builder.redirect_limit(None)
}
```

**Review by keelerm84 — APPROVED** [review]
